kong/clustering/wrpc_data_plane.lua

```diff
@@ -43,90 +50,90 @@ function _M.new(conf, cert, cert_key)
-          -- because `:count` will never yield, end result is that the semaphore
-          -- count is guaranteed to not exceed 1
-          peer.config_semaphore:post()
-        end
+local function init_config_service(service)
+  service:import("kong.services.config.v1.config")
+  service:set_handler("ConfigService.SyncConfig", function(peer, data)
+    -- yield between steps to prevent long delay
+    if peer.config_semaphore then
+      local json_config = assert(inflate_gzip(data.config))
+      yield()
+      peer.config_obj.next_config = assert(cjson_decode(json_config))
+      yield()
+
+      peer.config_obj.next_hash = data.config_hash
+      peer.config_obj.next_hashes = data.hashes
+      peer.config_obj.next_config_version = tonumber(data.version)
+      if peer.config_semaphore:count() <= 0 then
+        -- the following line always executes immediately after the `if` check
+        -- because `:count` will never yield, end result is that the semaphore
+        -- count is guaranteed to not exceed 1
+        peer.config_semaphore:post()
       end
-      return { accepted = true }
-    end)
-  end
-
-  return wrpc_config_service
+    end
+    return { accepted = true }
+  end)
 end
 
-function _M:communicate(premature)
-  if premature then
-    -- worker wants to exit
-    return
+local wrpc_services
+local function get_services()
+  if not wrpc_services then
+    wrpc_services = wrpc_proto.new()
+    init_negotiation_client(wrpc_services)
+    init_config_service(wrpc_services)
   end
 
-  local conf = self.conf
+  return wrpc_services
+end
+
+local function communicate_impl(dp)
+  local conf = dp.conf
 
   local log_suffix = " [" .. conf.cluster_control_plane .. "]"
-  local reconnection_delay = math.random(5, 10)
 
   local c, uri, err = clustering_utils.connect_cp(
-                        "/v1/wrpc", conf, self.cert, self.cert_key,
+                        "/v1/wrpc", conf, dp.cert, dp.cert_key,
                         "wrpc.konghq.com")
   if not c then
-    ngx_log(ngx_ERR, _log_prefix, "connection to control plane ", uri, " broken: ", err,
-                 " (retrying after ", reconnection_delay, " seconds)", log_suffix)
-
-    assert(ngx.timer.at(reconnection_delay, function(premature)
-      self:communicate(premature)
-    end))
-    return
+    error("connection to control plane " .. uri .." broken: " .. err)
   end
 
   local config_semaphore = semaphore.new(0)
-  local peer = wrpc.new_peer(c, get_config_service(), { channel = self.DPCP_CHANNEL_NAME })
+  local peer = wrpc.new_peer(c, get_services(), { channel = dp.DPCP_CHANNEL_NAME })
 
   peer.config_semaphore = config_semaphore
-  peer.config_obj = self
+  peer.config_obj = dp
   peer:spawn_threads()
 
   do
-    local resp, err = peer:call_async("ConfigService.ReportMetadata", { plugins = self.plugins_list })
+    local ok, err = negotiate(peer)
+    if not ok then
+      error(err)
+    end
+  end
+
+  do
```

Comment:
Now I think this `do ... end` is unnecessary.

Reply:
This is to limit the scope of variables.